components/instance-creator/internal/client/client.go

```diff
@@ -0,0 +1,446 @@
+// Example call, delete after initial usage of the client:
+//
+// DeleteResource(ctx, region, subaccountID, types.ServiceInstance{ID: id}, types.ServiceInstanceArguments{})
+func (c *client) DeleteResource(ctx context.Context, region, subaccountID string, resource resources.Resource, resourceArgs resources.ResourceArguments) error {
+	resourcePath := resourceArgs.GetURLPath() + fmt.Sprintf("/%s", resource.GetResourceID())
+	strURL, err := buildURL(c.cfg.InstanceSMURLPath, resourcePath, SubaccountKey, subaccountID)
+	if err != nil {
+		return errors.Wrapf(err, "while building %s URL", resource.GetResourceType())
+	}
+
+	req, err := http.NewRequest(http.MethodDelete, strURL, nil)
+	if err != nil {
+		return err
+	}
+
+	log.C(ctx).Infof("Deleting %s with ID: %q and subaccount: %q", resource.GetResourceType(), resource.GetResourceID(), subaccountID)
+	caller, err := c.callerProvider.GetCaller(c.cfg, region)
+	if err != nil {
+		return errors.Errorf("error while getting caller for region: %s", region)
+	}
+
+	resp, err := caller.Call(req)
+	if err != nil {
+		return err
+	}
+	defer closeResponseBody(ctx, resp)
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return errors.Errorf("failed to read response body from %s deletion request: %v", resource.GetResourceType(), err)
+	}
+
+	if resp.StatusCode != http.StatusOK && resp.StatusCode != http.StatusAccepted {
+		return errors.Errorf("failed to delete %s, status: %d, body: %s", resource.GetResourceType(), resp.StatusCode, body)
+	}
+
+	if resp.StatusCode == http.StatusAccepted {
+		log.C(ctx).Infof("Handle asynchronous %s deletion...", resource.GetResourceType())
+		_, err := c.executeAsyncRequest(ctx, resp, caller, subaccountID, false)
+		if err != nil {
+			return errors.Wrapf(err, "while deleting %s with ID: %q and subaccount: %q", resource.GetResourceType(), resource.GetResourceID(), subaccountID)
+		}
+		return nil
+	}
+
+	log.C(ctx).Infof("Successfully deleted %s with ID: %q and subaccount: %q synchronously", resource.GetResourceType(), resource.GetResourceID(), subaccountID)
+
+	return nil
+}
+
+// DeleteMultipleResources deletes multiple resources from SM by some criteria
+// Example call, delete after initial usage of the client:
+//
+// DeleteMultipleResources(ctx, region, subaccountID, &types.ServiceKeys{}, types.ServiceKeyArguments{ServiceInstanceID: id})
+func (c *client) DeleteMultipleResources(ctx context.Context, region, subaccountID string, resources resources.Resources, resourceArgs resources.ResourceArguments) error {
+	caller, err := c.callerProvider.GetCaller(c.cfg, region)
+	if err != nil {
+		return errors.Errorf("error while getting caller for region: %s", region)
+	}
+
+	resourceIDs, err := c.retrieveMultipleResources(ctx, caller, subaccountID, resources, resourceArgs)
+	if err != nil {
+		return err
+	}
+
+	for _, resourceID := range resourceIDs {
+		resourcePath := resourceArgs.GetURLPath() + fmt.Sprintf("/%s", resourceID)
```

Comment:
I don't think it will work because we have to call `DeleteResource()` with specific implementation of the `ResourceMatchParameters` and `Resource` interfaces but the `DeleteMultipleResources()` is generic so we don't have the knowledge what resource is being deleted.